## Laboratorium 3

### Opis
Celem laboratorium jest klasyfikacja liści za pomocą metody *k* najbliższych sąsiadów - *k*NN (*k Nearest Neighbors*) z wykorzystaniem normalizacji.


### Zbiór danych

Zbiór danych znajduje się w katalogu `dataset/*`. Jest to zmodyfikowany zbiór danych znajdujący się pod adresem: <https://archive.ics.uci.edu/ml/datasets/leaf>.

### Przesyłanie zadań

Wszystkie pliki należy spakować archiwizatorem **zip** i przesłać za pośrednictwem platformy WIKAMP. Poniżej oczekiwana zawartość archiwum:

```
+-- 📂 [IMIE.NAZWISKO].zip
    +-- 📜 Lab03.ipynb
    +-- 📂 dataset
        +-- 📜 dataset.npz
        +-- 📜 ReadMe.pdf
```



### Zadanie 1

* Wybierz 10 dowolnych gatunków liści (będziesz je używać w kolejnych zadaniach).
* Przeprowadź klasyfikację za pomocą klasyfikatora [*k*NN](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html).
* Znajdź optymalne *k*.
* Wyświetl najwyższy wynik klasyfikacji (*accuracy*).


In [119]:
import numpy as np

with open('./dataset/dataset.npz', 'rb') as f:
    data = np.load(f)   
    train, test = data['train'], data['test']
train = train[(train[:,0] >= 3) & (train[:,0] <= 12)]
train_t = train[:,0]
train = train[:, 2:]
test = test[(test[:,0] >= 3) & (test[:,0] <= 12)]
test_t = test[:,0]
test = test[:, 2:]
from sklearn.neighbors import KNeighborsClassifier
max_score = 0
best_k = 1
for k in range(1,50):
    neigh = KNeighborsClassifier(n_neighbors=k)
    neigh.fit(train, train_t)
    score = neigh.score(test, test_t)
    if score > max_score:
        max_score = score
        best_k = k
print(f"best K={best_k}, best score={max_score}")

best K=8, best score=0.7540106951871658


### Zadanie 2
* Znormalizuj dane w zakresie 0-1. Możesz do tego celu wykorzystać gotową funkcję z biblioteki scikit-learn https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html.
>    **Ważne: współczynniki powinny być obliczone na zbiorze treningowym i te same powinny zostać zastosowane do normalizacji zbioru testowego.**
* Ponownie wykonaj klasyfikację z tymi samymi parametrami (co w zadaniu 1), ale na danych znormalizowanych.
* Wyświetl najwyższy wynik klasyfikacji (*accuracy*).


In [121]:
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import StandardScaler

scaler = MinMaxScaler()
train = scaler.fit_transform(train)
test = scaler.transform(test)
max_score = 0
best_k = 1
for k in range(1,50):
    neigh_2 = KNeighborsClassifier(n_neighbors=k)
    neigh_2.fit(train, train_t)
    score = neigh_2.score(test, test_t)
    if score > max_score:
        max_score = score
        best_k = k
print(f"best K={best_k}, best score={max_score}")

best K=31, best score=0.8449197860962567


### Zadanie 3

Napisz wnioski

Dla przypadku nie uwzględniającego normalizacji, optymalne k z zakresu 1-50 to 8 a accuracy 0.75 natomiast po normalizacji wynik ten wzrósł o 9% do poziomu 0.84 dla optymalnego k = 31.\
Przyczyn polepszenia wyniku możemy szukać w naturze normalizacji, która sprowadza cechy do przedziału (0-1) dzięki czemu żadna z cech nie jest faworyzowana ze względu na wysoką jej wartość.